parser/prometheus_test.go

```diff
@@ -0,0 +1,129 @@
+/*
+ * skogul, Prometheus parser test
+ *
+ * Copyright (c) 2022 Telenor Norge AS
+ * Author(s):
+ *  - Roshini Narasimha Raghavan <roshiragavi@gmail.com>
+ *  - Kristian Lyngstøl <kly@kly.no>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+package parser_test
+
+import (
+	"os"
+	"testing"
+
+	"github.com/telenornms/skogul"
+	"github.com/telenornms/skogul/parser"
+)
+
+func TestPrometheus(t *testing.T) {
+	b, err := os.ReadFile("./testdata/prometheus_testdata")
+
+	if err != nil {
+		t.Errorf("Failed to read test data file: %v", err)
+		return
+	}
+	// convert byte array to json
+	p := parser.Prometheus{}
+
+	container, err := p.Parse(b)
+
+	if err != nil {
+		t.Logf("Error while parsing the data. %v", err)
+		t.FailNow()
+	}
+
+	//expectedTime := 1608520832877
+	metricKey1 := "dialer_name"
+	metricValue1 := "federate"
+	metricKey2 := "instance"
+	metricValue2 := "localhost:9090"
+	metricKey3 := "job"
+	metricValue3 := "prometheus"
+	var data1Value, data2Value, data3Value float64
+	data1Value = 1
+	data1Key := "net_conntrack_dialer_conn_attempted_total"
+	data2Key := "conntrack_dialer_conn_attempted_total"
+	data2Value = 1
+	data3Key := "dialer_conn_attempted_total"
+	data3Value = 1
+
+	if container == nil || container.Metrics == nil || len(container.Metrics) == 0 {
+		t.Logf("Expected parsed prometheus to return a container with at least 1 metric. Container: %v", container.Describe())
+		t.FailNow()
+	}
+	if len(container.Metrics) != 3 {
+		t.Errorf("Expected exactly 3 metrics, got %d", len(container.Metrics))
+	}
+
+	var m1 *skogul.Metric
+	var m2 *skogul.Metric
+	var m3 *skogul.Metric
+	for _, m := range container.Metrics {
+		if m.Data["net_conntrack_dialer_conn_attempted_total"] != nil {
+			m1 = m
+		}
+		if m.Data["conntrack_dialer_conn_attempted_total"] != nil {
+			m2 = m
+		}
+		if m.Data["dialer_conn_attempted_total"] != nil {
+			m3 = m
+		}
+	}
+
+	if m1 == nil || m2 == nil || m3 == nil {
+		t.Errorf("missing metrics?")
```

Comment:
This comment goes for the following test cases as well.

It's nice to explain what was expected and exactly what failed, so that when/if these tests fail some time in the future, it's easy to understand what was expected and what was received. The test cases for `mN.Data[dataNKey]` are well created, so if these for general metrics & metrics metadata could be updated to look like them, that'd be great.